wxPython not supported

The website mentions support for wxPython. But none of the examples can be made to work.

The fundamental error when trying to add a mayavi control to a wxPython window appears to be:

```
    sizer.Add(self.control, 0, wx.ALL, 10)
TypeError: Sizer.Add(): arguments did not match any overloaded call:
  overload 1: argument 1 has unexpected type 'QWidget'
```

Searching in my email I notice I participated in a discussion about this as far back as 2020: https://github.com/enthought/mayavi/issues/726 but that issue appears to have been removed.

Is there any interest now in adding support for wxPython back? 

If there is interest in adding this support, I can potentially help. I am not familiar with the mayavi implementation, but I am a regularcontributor to wxPython, and I have contributed to the integration between matplotlib and wxPython. (e.g. https://github.com/matplotlib/matplotlib/pull/26710)

Thank you. Tested in a fresh venv with both `wxPython` (4.3.1) and a Qt binding (PyQt5 + PySide6) installed, since that's what I have. Two separate things going on here:

**1. Toolkit auto-detection** — with a Qt binding also installed, ETS defaults to Qt, so the wx examples try to embed a `QWidget` in a wx container and blow up with the original error. `ETS_TOOLKIT=wx` fixes that, as you suspected.

**2. A real bug in the example itself**, which only surfaces once the toolkit is actually wx. With `ETS_TOOLKIT=wx` set, `wx_mayavi_embed_in_notebook.py` still fails, but with a different, legitimate error:

```
wx._core.wxAssertionError: C++ assertion "page->GetParent() == this" failed at .../auibook.cpp(2335) in InsertPage(): page must be a child of the notebook
```

Cause: [`examples/mayavi/interactive/wx_mayavi_embed_in_notebook.py`](https://github.com/enthought/mayavi/blob/main/examples/mayavi/interactive/wx_mayavi_embed_in_notebook.py) lines 60 and 68 call

```python
self.control = self.mayavi_view.edit_traits(
                parent=self,
                kind='subpanel').control
```

passing `parent=self` (the `wx.Frame`) instead of `parent=self.notebook`. Under Qt this is harmless because a widget gets silently reparented when added to a layout, but wx's `AuiNotebook.AddPage` requires the page to already be a child of the notebook, so it raises.

Fix is to parent the controls to the notebook instead of the frame:

```python
self.control = self.mayavi_view.edit_traits(
                parent=self.notebook,
                kind='subpanel').control
...
self.control2 = self.mayavi_view2.edit_traits(
                parent=self.notebook,
                kind='subpanel').control
```

With both `ETS_TOOLKIT=wx` and the parenting fix in `wx_mayavi_embed_in_notebook.py`, the example runs cleanly for me — window opens with both tabs, no crash, and the 3D object is displayed. The other example `wx_embedding.py` also runs and the 3D object is displayed.

--- a/docs/source/mayavi/auto/examples.rst
+++ b/docs/source/mayavi/auto/examples.rst
@@ -313,12 +313,12 @@ applications.
 
    example_mlab_visual.rst
    example_mlab_traits_ui.rst
-   example_wx_embedding.rst
    example_multiple_engines.rst
+   example_wx_embedding.rst
    example_multiple_mlab_scene_models.rst
    example_superquad_with_gui.rst
-   example_wx_mayavi_embed_in_notebook.rst
    example_mlab_interactive_dialog.rst
+   example_wx_mayavi_embed_in_notebook.rst
    example_subclassing_mayavi_application.rst
    example_mayavi_traits_ui.rst
    example_compute_in_thread.rst
@@ -342,11 +342,11 @@ applications.
     :width: 150
 
             
-.. |0102| image:: ../generated_images/example_wx_embedding.jpg
+.. |0102| image:: ../generated_images/example_multiple_engines.jpg
     :width: 150
 
             
-.. |0103| image:: ../generated_images/example_multiple_engines.jpg
+.. |0103| image:: ../generated_images/example_wx_embedding.jpg
     :width: 150
 
             
@@ -358,11 +358,11 @@ applications.
     :width: 150
 
             
-.. |0106| image:: ../generated_images/example_wx_mayavi_embed_in_notebook.jpg
+.. |0106| image:: ../generated_images/example_mlab_interactive_dialog.jpg
     :width: 150
 
             
-.. |0107| image:: ../generated_images/example_mlab_interactive_dialog.jpg
+.. |0107| image:: ../generated_images/example_wx_mayavi_embed_in_notebook.jpg
     :width: 150
 
             
@@ -418,13 +418,13 @@ applications.
          A simple example of how to use
          mayavi.mlab inside a traits UI dialog.
 
-|0102|  :ref:`example_wx_embedding`
-         This example shows to embed a Mayavi
-         view in a wx frame.
-
-|0103|  :ref:`example_multiple_engines`
+|0102|  :ref:`example_multiple_engines`
          An example to show how you can have
          multiple engines in one application.
+
+|0103|  :ref:`example_wx_embedding`
+         This example shows to embed a Mayavi
+         view in a wx frame.
 
 |0104|  :ref:`example_multiple_mlab_scene_models`
          Example showing a dialog with multiple
@@ -438,13 +438,13 @@ applications.
          parametrised by two parameters, *alpha*
          and *beta*.
 
-|0106|  :ref:`example_wx_mayavi_embed_in_notebook`
-         This example show how to embedded Mayavi
-         in a wx notebook.
-
-|0107|  :ref:`example_mlab_interactive_dialog`
+|0106|  :ref:`example_mlab_interactive_dialog`
          An example of how to modify the data
          visualized  via an interactive dialog.
+
+|0107|  :ref:`example_wx_mayavi_embed_in_notebook`
+         This example show how to embedded Mayavi
+         in a wx notebook.
 
 |0108|  :ref:`example_subclassing_mayavi_application`
          This script demonstrates how one can

--- a/docs/source/mayavi/auto/examples.rst
+++ b/docs/source/mayavi/auto/examples.rst
@@ -314,16 +314,16 @@ applications.
    example_mlab_visual.rst
    example_mlab_traits_ui.rst
    example_multiple_engines.rst
-   example_wx_embedding.rst
    example_multiple_mlab_scene_models.rst
    example_superquad_with_gui.rst
    example_mlab_interactive_dialog.rst
-   example_wx_mayavi_embed_in_notebook.rst
+   example_wx_embedding.rst
    example_subclassing_mayavi_application.rst
    example_mayavi_traits_ui.rst
    example_compute_in_thread.rst
    example_qt_embedding.rst
    example_adjust_cropping_extents.rst
+   example_wx_mayavi_embed_in_notebook.rst
    example_poll_file.rst
    example_lorenz_ui.rst
    example_volume_slicer.rst
@@ -346,44 +346,44 @@ applications.
     :width: 150
 
             
-.. |0103| image:: ../generated_images/example_wx_embedding.jpg
+.. |0103| image:: ../generated_images/example_multiple_mlab_scene_models.jpg
     :width: 150
 
             
-.. |0104| image:: ../generated_images/example_multiple_mlab_scene_models.jpg
+.. |0104| image:: ../generated_images/example_superquad_with_gui.jpg
     :width: 150
 
             
-.. |0105| image:: ../generated_images/example_superquad_with_gui.jpg
+.. |0105| image:: ../generated_images/example_mlab_interactive_dialog.jpg
     :width: 150
 
             
-.. |0106| image:: ../generated_images/example_mlab_interactive_dialog.jpg
+.. |0106| image:: ../generated_images/example_wx_embedding.jpg
     :width: 150
 
             
-.. |0107| image:: ../generated_images/example_wx_mayavi_embed_in_notebook.jpg
-    :width: 150
-
-            
-.. |0108| raw:: html
+.. |0107| raw:: html
 
     <br/>
 
             
-.. |0109| image:: ../generated_images/example_mayavi_traits_ui.jpg
+.. |0108| image:: ../generated_images/example_mayavi_traits_ui.jpg
     :width: 150
 
             
-.. |0110| image:: ../generated_images/example_compute_in_thread.jpg
+.. |0109| image:: ../generated_images/example_compute_in_thread.jpg
     :width: 150
 
             
-.. |0111| image:: ../generated_images/example_qt_embedding.jpg
+.. |0110| image:: ../generated_images/example_qt_embedding.jpg
     :width: 150
 
             
-.. |0112| image:: ../generated_images/example_adjust_cropping_extents.jpg
+.. |0111| image:: ../generated_images/example_adjust_cropping_extents.jpg
+    :width: 150
+
+            
+.. |0112| image:: ../generated_images/example_wx_mayavi_embed_in_notebook.jpg
     :width: 150
 
             
@@ -422,15 +422,11 @@ applications.
          An example to show how you can have
          multiple engines in one application.
 
-|0103|  :ref:`example_wx_embedding`
-         This example shows to embed a Mayavi
-         view in a wx frame.
-
-|0104|  :ref:`example_multiple_mlab_scene_models`
+|0103|  :ref:`example_multiple_mlab_scene_models`
          Example showing a dialog with multiple
          embedded scenes.
 
-|0105|  :ref:`example_superquad_with_gui`
+|0104|  :ref:`example_superquad_with_gui`
          This example uses MayaVi to show the
          evolution of a superquadric
          (http://en.wikipedia.org/wiki/Superquadrics),
@@ -438,27 +434,27 @@ applications.
          parametrised by two parameters, *alpha*
          and *beta*.
 
-|0106|  :ref:`example_mlab_interactive_dialog`
+|0105|  :ref:`example_mlab_interactive_dialog`
          An example of how to modify the data
          visualized  via an interactive dialog.
 
-|0107|  :ref:`example_wx_mayavi_embed_in_notebook`
-         This example show how to embedded Mayavi
-         in a wx notebook.
+|0106|  :ref:`example_wx_embedding`
+         This example shows to embed a Mayavi
+         view in a wx frame.
 
-|0108|  :ref:`example_subclassing_mayavi_application`
+|0107|  :ref:`example_subclassing_mayavi_application`
          This script demonstrates how one can
          script the Mayavi application by
          subclassing the application, create a
          new scene and create a few simple
          modules.
 
-|0109|  :ref:`example_mayavi_traits_ui`
+|0108|  :ref:`example_mayavi_traits_ui`
          An example of how to create a UI similar
          to the complete Mayavi application
          inside a Traits UI view.
 
-|0110|  :ref:`example_compute_in_thread`
+|0109|  :ref:`example_compute_in_thread`
          This script demonstrates how one can do
          a computation in another thread and
          update the mayavi pipeline. It also
@@ -466,14 +462,18 @@ applications.
          and visualize it as image data using a
          few modules.
 
-|0111|  :ref:`example_qt_embedding`
+|0110|  :ref:`example_qt_embedding`
          This example demonstrates using Mayavi
          as a component of a large Qt
          application.
 
-|0112|  :ref:`example_adjust_cropping_extents`
+|0111|  :ref:`example_adjust_cropping_extents`
          A custom dialog to adjust the parameters
          of a GeometryFilter to crop data points.
+
+|0112|  :ref:`example_wx_mayavi_embed_in_notebook`
+         This example show how to embedded Mayavi
+         in a wx notebook.
 
 |0113|  :ref:`example_poll_file`
          A simple script that polls a data file

--- a/docs/source/mayavi/auto/mayavi_traits_ui.py
+++ b/docs/source/mayavi/auto/mayavi_traits_ui.py
@@ -57,11 +57,11 @@ class Mayavi(HasTraits):
                                     editor=SceneEditor(),
                                     show_label=False,
                                     resizable=True,
-                                    height=500,
-                                    width=500),
+                                    ),
                         ),
                 resizable=True,
-                scrollable=True
+                scrollable=True,
+                id='mayavi.examples.interactive.mayavi_traits_ui'
                 )
 
     def __init__(self, **traits):

--- a/docs/source/mayavi/auto/wx_embedding.py
+++ b/docs/source/mayavi/auto/wx_embedding.py
@@ -15,6 +15,11 @@ In this example, the wx part is very simple. See
 embedding of Mayavi scenes in Wx applications.
 """
 
+# The toolkit is a per-process choice, and pyface prefers Qt when both are
+# installed, so a wx application has to claim it before anything reads it.
+from traits.etsconfig.api import ETSConfig
+ETSConfig.toolkit = 'wx'
+
 from numpy import ogrid, sin
 
 from traits.api import HasTraits, Instance

--- a/docs/source/mayavi/auto/wx_embedding.py
+++ b/docs/source/mayavi/auto/wx_embedding.py
@@ -53,6 +53,16 @@ class MayaviView(HasTraits):
 # Wx Code
 import wx
 
+
+def pop_event_handlers(window):
+    """ Undo the event handlers TraitsUI pushed onto a window tree.
+    """
+    while window.GetEventHandler() is not window:
+        window.PopEventHandler(True)
+    for child in window.GetChildren():
+        pop_event_handlers(child)
+
+
 class MainWindow(wx.Frame):
 
     def __init__(self, parent, id):
@@ -63,7 +73,17 @@ class MainWindow(wx.Frame):
         self.control = self.mayavi_view.edit_traits(
                         parent=self,
                         kind='subpanel').control
+        self.Bind(wx.EVT_CLOSE, self.on_close)
         self.Show(True)
+
+    def on_close(self, event):
+        # TraitsUI pushes wx event handlers onto the panel and its children and
+        # pops them only in ui.dispose(), which wx never gets to call on a
+        # frame that is simply closed; wx asserts on destroying a window that
+        # still has one.  Popping them by hand is enough -- disposing here
+        # instead schedules a second Destroy through wx.CallAfter and crashes.
+        pop_event_handlers(self.control)
+        event.Skip()
 
 app = wx.App(False)
 frame = MainWindow(None, wx.ID_ANY)

--- a/docs/source/mayavi/auto/wx_mayavi_embed_in_notebook.py
+++ b/docs/source/mayavi/auto/wx_mayavi_embed_in_notebook.py
@@ -49,6 +49,16 @@ class MayaviView(HasTraits):
 import wx
 import wx.aui
 
+
+def pop_event_handlers(window):
+    """ Undo the event handlers TraitsUI pushed onto a window tree.
+    """
+    while window.GetEventHandler() is not window:
+        window.PopEventHandler(True)
+    for child in window.GetChildren():
+        pop_event_handlers(child)
+
+
 class MainWindow(wx.Frame):
 
     def __init__(self, parent, id):
@@ -78,7 +88,21 @@ class MainWindow(wx.Frame):
         sizer.Add(self.notebook,1, wx.EXPAND)
         self.SetSizer(sizer)
 
+        self.Bind(wx.EVT_CLOSE, self.on_close)
         self.Show(True)
+
+    def on_close(self, event):
+        # TraitsUI pushes wx event handlers onto each panel and its children and
+        # pops them only in ui.dispose(), which wx never gets to call on a frame
+        # that is simply closed; wx asserts on destroying a window that still
+        # has one.  Popping them by hand is enough -- disposing here instead
+        # schedules a second Destroy through wx.CallAfter, and the notebook then
+        # crashes walking pages that are already gone.
+        # only the TraitsUI panels: the notebook pushes handlers of its own,
+        # and popping those aborts the interpreter on close
+        pop_event_handlers(self.control)
+        pop_event_handlers(self.control2)
+        event.Skip()
 
 if __name__ == '__main__':
     app = wx.App(False)

--- a/docs/source/mayavi/auto/wx_mayavi_embed_in_notebook.py
+++ b/docs/source/mayavi/auto/wx_mayavi_embed_in_notebook.py
@@ -70,9 +70,11 @@ class MainWindow(wx.Frame):
         self.mayavi_view = MayaviView()
 
         # The edit_traits method opens a first view of our 'MayaviView'
-        # object
+        # object.  A notebook page has to be a child of the notebook itself,
+        # not of the frame -- unlike a Qt layout, wx containers do not adopt
+        # what you put in them.
         self.control = self.mayavi_view.edit_traits(
-                        parent=self,
+                        parent=self.notebook,
                         kind='subpanel').control
         self.notebook.AddPage(page=self.control, caption='Display 1')
 
@@ -80,7 +82,7 @@ class MainWindow(wx.Frame):
 
         # The second call to edit_traits opens a second view
         self.control2 = self.mayavi_view2.edit_traits(
-                        parent=self,
+                        parent=self.notebook,
                         kind='subpanel').control
         self.notebook.AddPage(page=self.control2, caption='Display 2')
 

--- a/docs/source/mayavi/auto/wx_mayavi_embed_in_notebook.py
+++ b/docs/source/mayavi/auto/wx_mayavi_embed_in_notebook.py
@@ -9,6 +9,11 @@ In this example, we embed one single Mayavi scene in a Wx notebook, with
 2 tabs, each one of them hosting a different view of the scene.
 """
 
+# The toolkit is a per-process choice, and pyface prefers Qt when both are
+# installed, so a wx application has to claim it before anything reads it.
+from traits.etsconfig.api import ETSConfig
+ETSConfig.toolkit = 'wx'
+
 from numpy import ogrid, sin
 
 from traits.api import HasTraits, Instance

--- a/docs/source/mayavi/auto/wx_mayavi_embed_in_notebook.py
+++ b/docs/source/mayavi/auto/wx_mayavi_embed_in_notebook.py
@@ -104,6 +104,13 @@ class MainWindow(wx.Frame):
         # and popping those aborts the interpreter on close
         pop_event_handlers(self.control)
         pop_event_handlers(self.control2)
+        # AuiNotebook destroys its pages through the C++ virtual Destroy, and
+        # TraitsUIPanel.Destroy returns None where wx wants a bool; taking the
+        # pages off the notebook and destroying them from here avoids that call
+        while self.notebook.GetPageCount():
+            self.notebook.RemovePage(0)
+        self.control.Destroy()
+        self.control2.Destroy()
         event.Skip()
 
 if __name__ == '__main__':

--- a/docs/source/mayavi/building_applications.rst
+++ b/docs/source/mayavi/building_applications.rst
@@ -292,6 +292,13 @@ TraitsUI built -- calling ``dispose()`` here instead schedules a second
 ``Destroy`` through ``wx.CallAfter``, and popping over widgets of your
 own removes handlers wx installed for itself.  Both crash on close.
 
+Pass the container the panel is going into as ``parent``, not the frame
+around it: a widget handed to a Qt layout is reparented for you, while wx
+expects it to be a child of the container already.  Putting a Mayavi view
+in a notebook page therefore means ``parent=self.notebook`` -- with the
+frame as parent the page can come out mis-drawn, and recent wxWidgets
+assert on it outright rather than reparenting quietly as older ones did.
+
 Two examples of integrating Mayavi visualization with Wx applications are
 given:
 

--- a/docs/source/mayavi/building_applications.rst
+++ b/docs/source/mayavi/building_applications.rst
@@ -261,17 +261,36 @@ Using the `Visualization` class defined above::
 
     import wx
 
+    def pop_event_handlers(window):
+        while window.GetEventHandler() is not window:
+            window.PopEventHandler(True)
+        for child in window.GetChildren():
+            pop_event_handlers(child)
+
     class MainWindow(wx.Frame):
         def __init__(self, parent, id):
             wx.Frame.__init__(self, parent, id, 'Mayavi in Wx')
             self.visualization = Visualization()
             self.control = self.visualization.edit_traits(parent=self,
                                     kind='subpanel').control
+            self.Bind(wx.EVT_CLOSE, self.on_close)
             self.Show()
+
+        def on_close(self, event):
+            pop_event_handlers(self.control)
+            event.Skip()
 
     app = wx.App(False)
     frame = MainWindow(None, wx.ID_ANY)
     app.MainLoop()
+
+TraitsUI pushes wx event handlers onto the panel it builds and onto its
+children, and pops them only in ``ui.dispose()``, which never runs for a
+frame that is simply closed; wx then asserts on destroying a window that
+still has one.  Popping them as above is enough, and only over the panel
+TraitsUI built -- calling ``dispose()`` here instead schedules a second
+``Destroy`` through ``wx.CallAfter``, and popping over widgets of your
+own removes handlers wx installed for itself.  Both crash on close.
 
 Two examples of integrating Mayavi visualization with Wx applications are
 given:

--- a/docs/source/mayavi/building_applications.rst
+++ b/docs/source/mayavi/building_applications.rst
@@ -246,6 +246,17 @@ A complete, runnable, code based on the above comments is given in the
 Integrating in a WxPython application
 --------------------------------------
 
+Before defining the `Visualization` class, you should set the toolkit
+used by Traits to the Wx backend, as it otherwise prefers Qt when both
+are installed and `edit_traits` then hands you a Qt widget that no wx
+sizer will accept::
+
+    from traits.etsconfig.api import ETSConfig
+    ETSConfig.toolkit = 'wx'
+
+This has to happen before anything else reads the toolkit, so keep it at
+the top of the module, above the Traits and Mayavi imports.
+
 Using the `Visualization` class defined above::
 
     import wx

--- a/docs/source/render_examples.py
+++ b/docs/source/render_examples.py
@@ -481,7 +481,10 @@ def capture_in_subprocess(filename, image_file):
             'capture_one(%r, %r)\n' % (here, filename, image_file))
     short_name = os.path.splitext(os.path.basename(filename))[0]
     # the toolkit is per process, which is exactly what the isolation buys us:
-    # a wx example gets a wx child and the rest of the gallery stays on Qt
+    # a wx example gets a wx child and the rest of the gallery stays on Qt.
+    # The examples set ETSConfig.toolkit themselves too, but that cannot carry
+    # the child on its own: importing this module for capture_one resolves the
+    # toolkit to Qt through mlab, before the example is ever exec'd
     env = dict(os.environ)
     if is_wx_example(filename):
         env['ETS_TOOLKIT'] = 'wx'

--- a/examples/mayavi/interactive/wx_embedding.py
+++ b/examples/mayavi/interactive/wx_embedding.py
@@ -15,6 +15,11 @@ In this example, the wx part is very simple. See
 embedding of Mayavi scenes in Wx applications.
 """
 
+# The toolkit is a per-process choice, and pyface prefers Qt when both are
+# installed, so a wx application has to claim it before anything reads it.
+from traits.etsconfig.api import ETSConfig
+ETSConfig.toolkit = 'wx'
+
 from numpy import ogrid, sin
 
 from traits.api import HasTraits, Instance

--- a/examples/mayavi/interactive/wx_embedding.py
+++ b/examples/mayavi/interactive/wx_embedding.py
@@ -53,6 +53,16 @@ class MayaviView(HasTraits):
 # Wx Code
 import wx
 
+
+def pop_event_handlers(window):
+    """ Undo the event handlers TraitsUI pushed onto a window tree.
+    """
+    while window.GetEventHandler() is not window:
+        window.PopEventHandler(True)
+    for child in window.GetChildren():
+        pop_event_handlers(child)
+
+
 class MainWindow(wx.Frame):
 
     def __init__(self, parent, id):
@@ -63,7 +73,17 @@ class MainWindow(wx.Frame):
         self.control = self.mayavi_view.edit_traits(
                         parent=self,
                         kind='subpanel').control
+        self.Bind(wx.EVT_CLOSE, self.on_close)
         self.Show(True)
+
+    def on_close(self, event):
+        # TraitsUI pushes wx event handlers onto the panel and its children and
+        # pops them only in ui.dispose(), which wx never gets to call on a
+        # frame that is simply closed; wx asserts on destroying a window that
+        # still has one.  Popping them by hand is enough -- disposing here
+        # instead schedules a second Destroy through wx.CallAfter and crashes.
+        pop_event_handlers(self.control)
+        event.Skip()
 
 app = wx.App(False)
 frame = MainWindow(None, wx.ID_ANY)

--- a/examples/mayavi/interactive/wx_mayavi_embed_in_notebook.py
+++ b/examples/mayavi/interactive/wx_mayavi_embed_in_notebook.py
@@ -49,6 +49,16 @@ class MayaviView(HasTraits):
 import wx
 import wx.aui
 
+
+def pop_event_handlers(window):
+    """ Undo the event handlers TraitsUI pushed onto a window tree.
+    """
+    while window.GetEventHandler() is not window:
+        window.PopEventHandler(True)
+    for child in window.GetChildren():
+        pop_event_handlers(child)
+
+
 class MainWindow(wx.Frame):
 
     def __init__(self, parent, id):
@@ -78,7 +88,21 @@ class MainWindow(wx.Frame):
         sizer.Add(self.notebook,1, wx.EXPAND)
         self.SetSizer(sizer)
 
+        self.Bind(wx.EVT_CLOSE, self.on_close)
         self.Show(True)
+
+    def on_close(self, event):
+        # TraitsUI pushes wx event handlers onto each panel and its children and
+        # pops them only in ui.dispose(), which wx never gets to call on a frame
+        # that is simply closed; wx asserts on destroying a window that still
+        # has one.  Popping them by hand is enough -- disposing here instead
+        # schedules a second Destroy through wx.CallAfter, and the notebook then
+        # crashes walking pages that are already gone.
+        # only the TraitsUI panels: the notebook pushes handlers of its own,
+        # and popping those aborts the interpreter on close
+        pop_event_handlers(self.control)
+        pop_event_handlers(self.control2)
+        event.Skip()
 
 if __name__ == '__main__':
     app = wx.App(False)

--- a/examples/mayavi/interactive/wx_mayavi_embed_in_notebook.py
+++ b/examples/mayavi/interactive/wx_mayavi_embed_in_notebook.py
@@ -70,9 +70,11 @@ class MainWindow(wx.Frame):
         self.mayavi_view = MayaviView()
 
         # The edit_traits method opens a first view of our 'MayaviView'
-        # object
+        # object.  A notebook page has to be a child of the notebook itself,
+        # not of the frame -- unlike a Qt layout, wx containers do not adopt
+        # what you put in them.
         self.control = self.mayavi_view.edit_traits(
-                        parent=self,
+                        parent=self.notebook,
                         kind='subpanel').control
         self.notebook.AddPage(page=self.control, caption='Display 1')
 
@@ -80,7 +82,7 @@ class MainWindow(wx.Frame):
 
         # The second call to edit_traits opens a second view
         self.control2 = self.mayavi_view2.edit_traits(
-                        parent=self,
+                        parent=self.notebook,
                         kind='subpanel').control
         self.notebook.AddPage(page=self.control2, caption='Display 2')
 

--- a/examples/mayavi/interactive/wx_mayavi_embed_in_notebook.py
+++ b/examples/mayavi/interactive/wx_mayavi_embed_in_notebook.py
@@ -9,6 +9,11 @@ In this example, we embed one single Mayavi scene in a Wx notebook, with
 2 tabs, each one of them hosting a different view of the scene.
 """
 
+# The toolkit is a per-process choice, and pyface prefers Qt when both are
+# installed, so a wx application has to claim it before anything reads it.
+from traits.etsconfig.api import ETSConfig
+ETSConfig.toolkit = 'wx'
+
 from numpy import ogrid, sin
 
 from traits.api import HasTraits, Instance

--- a/examples/mayavi/interactive/wx_mayavi_embed_in_notebook.py
+++ b/examples/mayavi/interactive/wx_mayavi_embed_in_notebook.py
@@ -104,6 +104,13 @@ class MainWindow(wx.Frame):
         # and popping those aborts the interpreter on close
         pop_event_handlers(self.control)
         pop_event_handlers(self.control2)
+        # AuiNotebook destroys its pages through the C++ virtual Destroy, and
+        # TraitsUIPanel.Destroy returns None where wx wants a bool; taking the
+        # pages off the notebook and destroying them from here avoids that call
+        while self.notebook.GetPageCount():
+            self.notebook.RemovePage(0)
+        self.control.Destroy()
+        self.control2.Destroy()
         event.Skip()
 
 if __name__ == '__main__':

